src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs

```diff
@@ -2046,26 +2047,34 @@ static private void FlattenDeconstructVariables(ImmutableArray<DeconstructionVar
 
                 if (memberAccess.Kind != BoundKind.MethodGroup)
                 {
-                    Error(diagnostics, ErrorCode.ERR_MissingDeconstruct, receiverSyntax, receiver.Type);
-                    outPlaceholders = default(ImmutableArray<BoundDeconstructValuePlaceholder>);
-
-                    return BadExpression(receiverSyntax, receiver);
+                    return MissingDeconstruct(receiver, assignmentSyntax, numCheckedVariables, diagnostics, out outPlaceholders, receiver);
                 }
 
                 // After the overload resolution completes, the last step is to coerce the arguments with inferred types.
                 // That step returns placeholder (of correct type) instead of the outVar nodes that were passed in as arguments.
                 // So the generated invocation expression will contain placeholders instead of those outVar nodes.
                 // Those placeholders are also recorded in the outVar for easy access below, by the `SetInferredType` call on the outVar nodes.
+                bag = DiagnosticBag.GetInstance();
                 BoundExpression result = BindMethodGroupInvocation(
-                                            receiverSyntax, receiverSyntax, methodName, (BoundMethodGroup)memberAccess, analyzedArguments, diagnostics, queryClause: null,
+                                            receiverSyntax, receiverSyntax, methodName, (BoundMethodGroup)memberAccess, analyzedArguments, bag, queryClause: null,
                                             allowUnexpandedForm: true);
 
                 result.WasCompilerGenerated = true;
 
-                if (result.HasAnyErrors)
+                if (bag.HasAnyErrors())
+                {
+                    return MissingDeconstruct(receiver, assignmentSyntax, numCheckedVariables, diagnostics, out outPlaceholders, result);
+                }
+                else
+                {
+                    diagnostics.AddRange(bag);
+                }
+
+                var deconstructMethod = ((BoundCall)result).Method;
+                var parameters = deconstructMethod.IsExtensionMethod ? deconstructMethod.Parameters.Skip(1) : deconstructMethod.Parameters;
+                if (parameters.Any(p => p.RefKind != RefKind.Out && !p.IsThis))
```

Comment:
It might be easier to just do a small for loop here and start with 1 in case of extension methods. That would also avoid allocation since skip allocates. That is not a big deal though.